### PR TITLE
Allocate intermediate tensors based on allocation domain

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I don't know why we were allocating based on leaf domain (`alloc->shape()`), but I don't think this is correct.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/codegen.cpp [hunk 1]
```diff
@@ -230,7 +230,8 @@ class CudaKernelGenerator : private OptOutConstDispatch {
     for (auto allocate : kernel_summary.global_allocations) {
       TORCH_INTERNAL_ASSERT(allocate->buffer()->isA<TensorView>());
       const auto tv = allocate->buffer()->as<TensorView>();
-      const auto& alloc_domain = tv->getMaybeAllocationDomain();
+      const auto& alloc_domain =
+          TensorDomain::noReductions(tv->getMaybeAllocationDomain());
       const auto nDims = std::count_if(
```
Comment (naoyam) [inline csrc/codegen.cpp:235]: This can be just `alloc_domain.size()`.

**Comment by zasdfgbnm** [comment]
> That would allow omitting some of the non-divisible split checks.

What do you mean? Are you saying that using leaf domain as allocation will save non-divisible split checks? If this is what you mean, then I agree on this part. But IIUC, currently we are allocating based on leaf domain, but indexing into allocation domain? I think this would work only if the allocation is contiguous. Assume we allocated a discontiguous tensor, and correctly set the contiguity to false, we would still get illegal memory access.

File: csrc/codegen.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/codegen.cpp]: fixed

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
I think this is not a performance issue, but a correctness issue? In codegen, they are generated with the same dimensionality as the allocation domain, then we will have code like `T1.stride[2]`, however, in the allocation, we will not fill the correct stride, because the stride that the executor fill is the stride of leaf domains. For example, if the leaf domain only has 1 dimension, then `T1.stride[2]` is clearly an out of bound access on constant memory.

**Comment by naoyam** [comment]
I'm not sure I understand what you said, but before the introduction of allocation domains, allocations of global memory tensors are done based on rfactor domains. (https://github.com/NVIDIA/Fuser/blob/main/csrc/lower_allocation.cpp#L160). The executor allocates and fills meta data of global memory tensors based on rfactor domains. 

For non-global tensors, allocations are based on leaf domains. That's to some extent natural as we want to only allocate domains right of computeAt positions. This also means if there's a non-divisible split, we would overallocate, and I remember at one point we used that property to omit predicates for non-divisible splits. Maybe this one relies on that https://github.com/NVIDIA/Fuser/blob/main/csrc/lower_predicate_elimination.cpp#L493

File: csrc/executor.cpp [hunk 2]
```diff
@@ -599,14 +599,14 @@ std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShape(
 // Infer the shape of an intemediate tensor using kir::Allocate
 std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShapeOfIntermediate(
     const TensorView* tv,
-    const kir::Allocate* alloc,
     ExpressionEvaluator& expr_eval) {
-  // Allocate should be provided for intermediates. We just need to
-  // grab a chunk of memory of the size dicatated by
-  // Allocate::shape().
-  TORCH_INTERNAL_ASSERT(alloc != nullptr);
+  auto alloc_dom = TensorDomain::noReductions(tv->getMaybeAllocationDomain());
```
Comment (naoyam) [inline csrc/executor.cpp:603]: Shouldn't this also ignore broadcast domains? See `getGlobalAllocationSizes` in lower_allocation.cpp

**Comment by zasdfgbnm** [comment]
Oh I was talking about https://github.com/NVIDIA/Fuser/blob/75da2061e56e8efdb0f13b7997714cae10eb164a/csrc/kernel_ir.cpp#L142-L145
which is using the leaf domain removing reductions.
Seems that this path is not used for global tensors?

**Comment by naoyam** [comment]
> Oh I was talking about
> 
> https://github.com/NVIDIA/Fuser/blob/75da2061e56e8efdb0f13b7997714cae10eb164a/csrc/kernel_ir.cpp#L142-L145
> 
> 
> which is using the leaf domain removing reductions.
> Seems that this path is not used for global tensors?

That seems to be used only for very limited cases, like allocating temporary work buffers for grid reductions. For tensors that are defined in original fusions, I believe they are all handled in `AllocationInserter` of lower_allocation.

File: csrc/executor.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/executor.cpp]: Removed broadcasts.

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by zasdfgbnm** [comment]
Oh, wait, I don't think this is correct. When generating index, for example, if the allocation domain is `[I{100}, b{200}, I{300}]`, then we will generate something like `i0 * stride[0] + i2 * stride[2]`. So we should not just get rid of broadcast. Instead, we should make the shape of broadcast `1`, regardless of its actually extent. Marking this as draft.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
@naoyam Do you want to take another look at this PR?

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**